src/20250114_AaveV3Ethereum_GSMsMigrationToGSM4626/AaveV3Ethereum_GSMsMigrationToGSM4626_20250114.sol

```diff
@@ -0,0 +1,252 @@
+    );
+    uint256 sharesUsdt = IERC4626StataToken(AaveV3EthereumAssets.USDT_STATA_TOKEN).deposit(
+      balanceUsdt,
+      address(this)
+    );
+
+    IERC20(AaveV3EthereumAssets.USDC_STATA_TOKEN).forceApprove(NEW_GSM_USDC, sharesUsdc);
+    IERC20(AaveV3EthereumAssets.USDT_STATA_TOKEN).forceApprove(NEW_GSM_USDT, sharesUsdt);
+
+    (, uint256 amountGhoUsdc) = IGsm(NEW_GSM_USDC).sellAsset(sharesUsdc, address(this));
+    (, uint256 amountGhoUsdt) = IGsm(NEW_GSM_USDT).sellAsset(sharesUsdt, address(this));
+
+    (, uint256 ghoUsdcNeeded) = IGhoToken(GhoEthereum.GHO_TOKEN).getFacilitatorBucket(
+      GhoEthereum.GSM_USDC
+    );
+    (, uint256 ghoUsdtNeeded) = IGhoToken(GhoEthereum.GHO_TOKEN).getFacilitatorBucket(
+      GhoEthereum.GSM_USDT
+    );
+
+    uint256 difference = ghoUsdcNeeded + ghoUsdtNeeded - amountGhoUsdc - amountGhoUsdt;
+    if (difference > 0) {
+      AaveV3Ethereum.COLLECTOR.transfer(GhoEthereum.GHO_TOKEN, address(this), difference);
+    }
+
+    IERC20(GhoEthereum.GHO_TOKEN).forceApprove(GhoEthereum.GSM_USDC, ghoUsdcNeeded);
+    IERC20(GhoEthereum.GHO_TOKEN).forceApprove(GhoEthereum.GSM_USDT, ghoUsdtNeeded);
+
+    IGsm(GhoEthereum.GSM_USDC).burnAfterSeize(ghoUsdcNeeded);
+    IGsm(GhoEthereum.GSM_USDT).burnAfterSeize(ghoUsdtNeeded);
+
+    IGsm(GhoEthereum.GSM_USDC).distributeFeesToTreasury();
+    IGsm(GhoEthereum.GSM_USDT).distributeFeesToTreasury();
+  }
+
+  function _revokeAccess() internal {
+    // Remove existing GSMs as GHO Facilitators
+    IGhoToken(GhoEthereum.GHO_TOKEN).removeFacilitator(GhoEthereum.GSM_USDC);
+    IGhoToken(GhoEthereum.GHO_TOKEN).removeFacilitator(GhoEthereum.GSM_USDT);
+
+    // Revoke existing GSMs
+    address[] memory revokedVaults = new address[](2);
+    revokedVaults[0] = GhoEthereum.GSM_USDC;
+    revokedVaults[1] = GhoEthereum.GSM_USDT;
+    IGhoBucketSteward(GhoEthereum.GHO_BUCKET_STEWARD).setControlledFacilitator(
+      revokedVaults,
+      false
+    );
+
+    // Remove existing GSMs from Registry
+    IGsmRegistry(GhoEthereum.GSM_REGISTRY).removeGsm(GhoEthereum.GSM_USDC);
+    IGsmRegistry(GhoEthereum.GSM_REGISTRY).removeGsm(GhoEthereum.GSM_USDT);
+
+    // Revoke Roles
+    IGsm(GhoEthereum.GSM_USDC).revokeRole(
+      SWAP_FREEZER_ROLE,
+      GhoEthereum.GSM_USDC_ORACLE_SWAP_FREEZER
+    );
+    IGsm(GhoEthereum.GSM_USDT).revokeRole(
+      SWAP_FREEZER_ROLE,
+      GhoEthereum.GSM_USDT_ORACLE_SWAP_FREEZER
+    );
+    IGsm(GhoEthereum.GSM_USDC).revokeRole(SWAP_FREEZER_ROLE, GovernanceV3Ethereum.EXECUTOR_LVL_1);
+    IGsm(GhoEthereum.GSM_USDT).revokeRole(SWAP_FREEZER_ROLE, GovernanceV3Ethereum.EXECUTOR_LVL_1);
+
+    // Cancel existing keepers
+    IAaveCLRobotOperator(MiscEthereum.AAVE_CL_ROBOT_OPERATOR).cancel(
+      EXISTING_ORACLE_SWAP_FREEZER_USDC
+    );
+    IAaveCLRobotOperator(MiscEthereum.AAVE_CL_ROBOT_OPERATOR).cancel(
+      EXISTING_ORACLE_SWAP_FREEZER_USDT
+    );
+
+    // Withdraw LINK from existing keepers permissionesly ~20 blocks after cancel
```

Comment:
how is this done?

Reply:
@miguelmtzinf it's a permissionless function that can be called by anyone later on, unfortunately cannot be done on the same TX as cancel. so i'll just go ahead and call it after execution.